### PR TITLE
[Concurrency] Import "did" delegate methods as @asyncHandler.

### PR DESCRIPTION
**Pull request by DougGregor**
Infer `@asyncHandler` on a protocol methods that follow the delegate
convention of reporting that something happened via a "did" method, so
long as they also meet the constraints for an `@asyncHandler` method in
Swift. This enables inference of `@asyncHandler` for witnesses of these
methods.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -4121,6 +4121,9 @@ ERROR(asynchandler_async,none,
 ERROR(asynchandler_inout_parameter,none,
       "'inout' parameter is not allowed in '@asyncHandler' function",
       ())
+ERROR(asynchandler_unsafe_pointer_parameter,none,
+      "'%0' parameter is not allowed in `@asyncHandler` function",
```
Comment (theblixguy) [inline include/swift/AST/DiagnosticsSema.def:4125]:
```suggestion
      "'%0' parameter is not allowed in '@asyncHandler' function",
```

Comment (DougGregor) [reply include/swift/AST/DiagnosticsSema.def:4125]: Ah, you're right. I considered banning it outright, then forgot to remove the diagnostic. Thank you!

**Comment by DougGregor** [comment]
@swift-ci smoke test

File: include/swift/AST/Decl.h [hunk 2]
```diff
@@ -5948,6 +5948,10 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
   /// Returns true if the function is an @asyncHandler.
   bool isAsyncHandler() const;
 
+  /// Returns true if the function if the signature matches the form of an
```
Comment (LucianoPAlmeida) [inline include/swift/AST/Decl.h:5951]: It the statement "if the function if the signature matches..." is correct? Or it should be something like "if the function signature matches..."?

Comment (DougGregor) [reply include/swift/AST/Decl.h:5951]: "If the function signature matches" is correct; will fix in a follow-up.

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Pull request merged**